bindings/ergo-lib-wasm/src/transaction.rs

```diff
@@ -153,6 +155,25 @@ pub struct UnsignedTransaction(chain::transaction::unsigned::UnsignedTransaction
 
 #[wasm_bindgen]
 impl UnsignedTransaction {
+    /// Create an UnsignedTransaction instance containing the given ContextExtension
+    pub fn with_context_ext(
```

Comment:
Updated, I'm still wrapping my head around rust ownership/mutability rules so it might be a bit off.

Thanks as always for the explanations!

Reply:
> In `with_*` methods we have a "convention" to return the original instance with "added" something passed as a parameter.

This does not sound right. I missed 2 key points - we **consume** the original and return an updated **copy** with "added/changed" something. However, this approach usually does not make good UX in JS land, since the "consumed" variable silently becomes null. So most of the time we end up with a mutating self. This works well if the method is infallible. If a method can return an error (like in this case) there is a good chance that JS caller may forget to handle it. So for cases that we want to "force" JS caller to handle error we use consuming self and return a new version with updated property with a detailed doc comment (see [`ErgoTree::with_constant`example](https://github.com/ergoplatform/sigma-rust/blob/86e012029ffab9590ef0898ff98f11bd30b8b004/bindings/ergo-lib-wasm/src/ergo_tree.rs#L59-L67) ).

Given how many points I missed :facepalm: in my first comment could you please change the code one more time?
Let's make it consume self and return an updated copy: `pub fn with_input_context_ext(self, input_id: &BoxId, ext: &ContextExtension) -> Result<UnsignedTransaction, JsValue>`

Reply:
Thanks again for the detailed explanation, very helpful :)

I've committed a version with mutable self as I wasn't sure how to pass self as immutable without clippy errors around redundant copies. I was trying something like this with the immutable self approach:

```rust
  pub fn with_input_context_ext(
      self,
      input_id: &BoxId,
      ext: &ContextExtension,
  ) -> Result<UnsignedTransaction, JsValue> {
      let mut cloned_self = self.clone();
      let mut input = cloned_self 
          .0
          .inputs
          .iter_mut()
          .find(|input| input.box_id == input_id.clone().into())
          .ok_or_else(|| JsValue::from_str("Box input id not found"))?;
      input.extension = ext.clone().into();

      Ok(cloned_self)
  }
```

Happy to make any more changes if what I've done is incorrect of course! 👍